### DISCUSSION TITLE
Allow for dynamic resolution of the storeDir for collectFile

### DISCUSSION THREAD

**Discussion by stevekm** [nextflow-io/nextflow#4461] [Q&A] [answered]
# New Feature Request

The [`collectFile`](https://nextflow.io/docs/latest/operator.html#collectfile) Channel operator has the ability to dynamically set the file `name` argument from within a closure like this;

```groovy
map_ch = Channel.of(
        ["id":"S1", "val":10],
        ["id":"S1", "val":20],
        ["id":"S2", "val":30],
        ["id":"S2", "val":40])

map_ch.collectFile(storeDir: "output", newLine: true) { meta ->
        def filename = "${meta.id}.txt"
        return [filename, meta.val.toString()]
    }

// output/S1.txt
// output/S2.txt
```

However it would be really helpful to be able to save files in this manner with dynamically set sub-directories as well. Instead of saving to 

```
output/S1.txt
output/S2.txt
```

I want to be able to save to 

```
output/S1/S1.txt
output/S2/S2.txt
```

It does not seem like its currently possible to do this. You cannot use any dynamic methods in the `storeDir` arg, and trying to put a subdir in the `name` arg seems to break often, especially with the `-resume` feature;

```groovy
map_ch.collectFile(storeDir: "${params.outdir}") { meta ->
        def filename = "${meta.id}/${meta.id}.txt"
        return [filename, meta.val.toString()]
    }
```

```
ERROR ~ No such file or directory: output/S1/S1.txt
```

Is there any way we could get a feature like this? It would help a lot with collection of files and metadata from process outputs that we want to keep associated with e.g. per-sample output subdirectories. The only other way I have been able to figure out so far for outputting files in per-sample subdirs is to [use a Nextflow process](https://github.com/nextflow-io/nextflow/issues/1515#issuecomment-1718876239) but that is not ideal if we can just do it with `collectFile` instead

**Comment by mribeirodantas** [comment, marked as answer]
You could create the folder from within your closure. I added `new File("output/${meta.id}").mkdirs()` to your code below and it works.

```groovy
map_ch = Channel.of(
        ["id":"S1", "val":10],
        ["id":"S1", "val":20],
        ["id":"S2", "val":30],
        ["id":"S2", "val":40])

map_ch.collectFile(storeDir: "output", newLine: true) { meta ->
        new File("output/${meta.id}").mkdirs()
        def filename = "${meta.id}/${meta.id}.txt"
        return [filename, meta.val.toString()]
    }
```

Folder structure:
```
tree output/
output/
├── S1
│   └── S1.txt
└── S2
    └── S2.txt

3 directories, 2 files
```
